File: app/src/main/java/org/antrack/app/service/Alarm.java
package org.antrack.app.service;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.SystemClock;

import org.antrack.app.C;

public class Alarm extends BroadcastReceiver {
    private static PendingIntent pIntent = null;

    @Override
    public void onReceive(Context context, Intent intent) {
        Intent sIntent = new Intent(context, MainService.class);
        sIntent.setAction(C.ACTION_ALARM);

        context.startService(sIntent);
    }

    public static void set(Context context, long time) {
        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(context, Alarm.class);
        pIntent = PendingIntent.getBroadcast(context, 0, intent, 0);

        am.setInexactRepeating(AlarmManager.ELAPSED_REALTIME_WAKEUP, SystemClock.elapsedRealtime(), time, pIntent);
    }

    public static void cancel(Context context) {
        if (pIntent != null) {
            AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
            am.cancel(pIntent);
        }
    }
}

File: app/src/main/java/org/antrack/app/service/Alarm.kt
package org.antrack.app.service

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.SystemClock
import org.antrack.app.App

class Alarm : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        CloudService.start(context, IntentActions.ACTION_ALARM)
    }

    companion object {
        private val am by lazy { App.context.getSystemService(Context.ALARM_SERVICE) as AlarmManager }

        private val pIntent by lazy {
            PendingIntent.getBroadcast(
                App.context,
                0,
                Intent(App.context, Alarm::class.java),
                PendingIntent.FLAG_IMMUTABLE
            )
        }

        fun set(time: Long) {
            am.setInexactRepeating(
                AlarmManager.ELAPSED_REALTIME_WAKEUP,
                SystemClock.elapsedRealtime(),
                time,
                pIntent
            )
        }

        fun cancel() {
            if (pIntent != null) {
                am.cancel(pIntent)
            }
        }
    }
}
